include/fiction/networks/views/input_ordering_view.hpp

```diff
@@ -0,0 +1,510 @@
+                                                this->set_visited(fin_inp, this->trav_id());
+                                            }
+                                            /* Mark FOs with already one related PI */
+                                            already_one_pi = true;
+                                            first_pi       = fin_inp;
+                                        }
+                                    }
+                                    else
+                                    {
+                                        if (inv_flag)
+                                        {
+                                            /* when a PI is related to a PI it is ranked third */
+                                            if (this->visited(n) != this->trav_id())
+                                            {
+                                                third_wait.push_back(n);
+                                                this->set_visited(n, this->trav_id());
+                                            }
+                                            if (this->visited(fin_inp) != this->trav_id())
+                                            {
+                                                third_wait.push_back(fin_inp);
+                                                this->set_visited(fin_inp, this->trav_id());
+                                            }
+                                        }
+                                        else
+                                        {
+                                            /* when a PI is related to a PI it is ranked third */
+                                            if (this->visited(fin_inp) != this->trav_id())
+                                            {
+                                                third_wait.push_back(fin_inp);
+                                                this->set_visited(fin_inp, this->trav_id());
+                                            }
+                                            if (this->visited(n) != this->trav_id())
+                                            {
+                                                third_wait.push_back(n);
+                                                this->set_visited(n, this->trav_id());
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        });
+                }
+            });
+    }
+
+    // private
+    Ntk               ntk;
+    std::vector<node> topo_order{};
+
+    std::vector<node> wait{};         // FOs related to 2 PIs
+    std::vector<node> second_wait{};  // FOs related to 1PI
+    std::vector<node> third_wait{};   // PI related to PI
+
+    bool fo_inv_flag     = false;
+    uint64_t  fo_inv_flag_num{0u};
+
+    uint32_t num_p;
+    uint32_t num_c{0u};
+    uint32_t num_r{0u};
```

Comment:
Please add docstrings.

Also, we usually put parameters first, and functions second in each scope (`public`, `protected`, `private`).